File: AquaTrack/src/main/java/com/aquatrack/service/BillingCycleService.java
package com.aquatrack.service;

import com.aquatrack.dto.billingcycle.BillingCycleResponseDto;
import com.aquatrack.dto.billingcycle.CreateBillingCycleRequestDto;
import com.aquatrack.dto.billingcycle.UpdateBillingCycleRequestDto;

import java.util.List;

public interface BillingCycleService {

    // ==========================================
    // Create
    // ==========================================

    BillingCycleResponseDto createBillingCycle(
            CreateBillingCycleRequestDto requestDto
    );

    // ==========================================
    // Read
    // ==========================================

    /**
     * Get all billing cycles accessible to the logged-in manager.
     */
    List<BillingCycleResponseDto> getAllBillingCycles();

    /**
     * Get billing cycle by ID.
     */
    BillingCycleResponseDto getBillingCycleById(
            Long billingCycleId
    );

    /**
     * Get billing cycles of a building.
     */
    List<BillingCycleResponseDto> getBillingCyclesByBuilding(
            Long buildingId
    );

    // ==========================================
    // Update
    // ==========================================

    BillingCycleResponseDto updateBillingCycle(
            Long billingCycleId,
            UpdateBillingCycleRequestDto requestDto
    );

    // ==========================================
    // Billing Status
    // ==========================================

    /**
     * Close an OPEN billing cycle.
     */
    BillingCycleResponseDto closeBillingCycle(
            Long billingCycleId
    );

    // ==========================================
    // Delete
    // ==========================================

    void deleteBillingCycle(
            Long billingCycleId
    );

}

File: Aquatrack-frontend/src/services/billingCycleService.js
import api from "../api/axios";

// ==========================================
// Create Billing Cycle
// ==========================================

export const createBillingCycle = async (data) => {

    const response = await api.post(
        "/manager/billing-cycles",
        data
    );

    return response.data;

};

// ==========================================
// Get All Billing Cycles
// ==========================================

export const getAllBillingCycles = async () => {

    const response = await api.get(
        "/manager/billing-cycles"
    );

    return response.data.data;

};

// ==========================================
// Get Billing Cycle By ID
// ==========================================

export const getBillingCycleById = async (billingCycleId) => {

    const response = await api.get(
        `/manager/billing-cycles/${billingCycleId}`
    );

    return response.data.data;

};

// ==========================================
// Get Billing Cycles By Building
// ==========================================

export const getBillingCyclesByBuilding = async (buildingId) => {

    const response = await api.get(
        `/manager/billing-cycles/building/${buildingId}`
    );

    return response.data.data;

};

// ==========================================
// Update Billing Cycle
// ==========================================

export const updateBillingCycle = async (
    billingCycleId,
    data
) => {

    const response = await api.put(
        `/manager/billing-cycles/${billingCycleId}`,
        data
    );

    return response.data;

};

// ==========================================
// Close Billing Cycle
// ==========================================

export const closeBillingCycle = async (billingCycleId) => {

    const response = await api.patch(
        `/manager/billing-cycles/${billingCycleId}/close`
    );

    return response.data;

};

// ==========================================
// Delete Billing Cycle
// ==========================================

export const deleteBillingCycle = async (billingCycleId) => {

    const response = await api.delete(
        `/manager/billing-cycles/${billingCycleId}`
    );

    return response.data;

};